### PR TITLE
Add compatible command group

### PR DESCRIPTION
**Pull request by Jimbo4350**
Related: #926 
# Changelog

```yaml
- description: |
    Introduce "compatible" command group
  type:
  - feature        # introduces a new feature
  - compatible     # the API has changed but is non-breaking
```

# Context

```
Usage: cardano-cli compatible 
                                ( shelley
                                | allegra
                                | mary
                                | alonzo
                                | babbage
                                | conway
                                )

  Limited backward compatible commands for testing only.

Available options:
  -h,--help                Show this help text

Available commands:
  shelley                  Shelley era commands
  allegra                  Allegra era commands
  mary                     Mary era commands
  alonzo                   Alonzo era commands
  babbage                  Babbage era commands
  conway                   Conway era commands

```
# How to trust this PR

Highlight important bits of the PR that will make the review faster. If there are commands the reviewer can run to observe the new behavior, describe them.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Self-reviewed the diff

### CODE REVIEW

File: cardano-cli/src/Cardano/CLI/Commands.hs [hunk 1]
```diff
@@ -23,7 +24,9 @@ data ClientCommand
   | AddressCommand AddressCmds
   | -- | Byron Related Commands
     ByronCommand ByronCommand
-  | -- | Era agnostic hashing commands
+  | -- | Backward compatible commands for testing only
+    CompatibaleCommands AnyCompatibleCommand
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Commands.hs]: SP

File: cardano-cli/src/Cardano/CLI/Compatible/Commands.hs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+data CompatibleCommand era
+  = CompatibleTransactionCmd (CompatibleTransactionCmds era)
+  | CompatibleGovernanceCmds (CompatibleGovernanceCmds era)
+
+renderCompatibleCommand :: CompatibleCommand era -> Text
+renderCompatibleCommand = \case
+  CompatibleTransactionCmd cmd -> renderCompatibleTransactionCmd cmd
+  CompatibleGovernanceCmds cmd -> renderCompatibleGovernanceCmds cmd
+
+pAnyCompatibleCommand :: EnvCli -> Parser AnyCompatibleCommand
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/Compatible/Commands.hs:46]: Parsers should go to `cardano-cli/src/Cardano/CLI/Compatible/Options/Common.hs` or something similar

File: cardano-cli/src/Cardano/CLI/Compatible/Governance.hs
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/Compatible/Governance.hs:1]: This file should be named
`cardano-cli/src/Cardano/CLI/Compatible/Options/Governance.hs`

File: cardano-cli/src/Cardano/CLI/Compatible/Transaction.hs
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/Compatible/Transaction.hs:1]: This file should be named `cardano-cli/src/Cardano/CLI/Compatible/Options/Transaction.hs`

File: cardano-cli/src/Cardano/CLI/Compatible/Transaction.hs [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TupleSections #-}
+
+module Cardano.CLI.Compatible.Transaction
+  ( CompatibleTransactionCmds (..)
+  , CompatibleTransactionError (..)
+  , pAllCompatibleTransactionCommands
+  , renderCompatibleTransactionCmd
+  , runCompatibleTransactionCmd
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/Compatible/Transaction.hs:12]: This module should be split into Run, Option and Command modules

**Review by carbolymer — COMMENTED** [review]
What's the plan with Eons and compatible commands? Are we planning to keep necessary eons for those commands?

**Review by carbolymer — APPROVED** [review]

File: cardano-cli/src/Cardano/CLI/Compatible/Commands.hs (hunk 2 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Compatible/Commands.hs:46]: I don't want to mirror the previous `EraBased` directory structure. A `Common` module is fine but this parser is not common i.e it's called in one place.

File: cardano-cli/src/Cardano/CLI/Compatible/Governance.hs
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Compatible/Governance.hs:1]: Same as above, I don't want to mirror the previous directory structure.

File: cardano-cli/src/Cardano/CLI/Compatible/Transaction.hs (hunk 3 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Compatible/Transaction.hs:12]: https://github.com/IntersectMBO/cardano-cli/pull/917#discussion_r1810738996

**Comment by Jimbo4350** [comment]
> What's the plan with Eons and compatible commands? Are we planning to keep necessary eons for those commands?

Currently those eons will be intended for use only with our compatibility commands and we should add this to the haddocks in `cardano-api`.  This will limit the scope of how much intra era compatibility code we have to write vs before where we had to write and maintain code for everything that changed between eras.

**Pull request merged**